### PR TITLE
Return miner uncles reward on total balances

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: store/transfer_processor.go [hunk 1]
```diff
@@ -88,12 +84,14 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
 		}
 	}
-	// From transfer events
+	// From transfer events (including miner and uncles)
 	for _, e := range events {
 		fromAddr := gethCommon.BytesToAddress(e.From)
 		if _, ok := seenAddrs[fromAddr]; !ok {
-			seenAddrs[fromAddr] = struct{}{}
-			addrs = append(addrs, e.From)
+			if !e.IsMinerRewardEvent() && !e.IsUncleRewardEvent() {
```
Comment (markya0616) [inline store/transfer_processor.go]: Suggest
```
if e.IsMinerRewardEvent() || e.IsUncleRewardEvent() {
    continue
}
fromAddr := gethCommon.BytesToAddress(e.From)
...
```

File: store/transfer_processor.go [hunk 2]
```diff
@@ -169,17 +167,19 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
-			continue
-		}
-		if contractsAddrs[model.ETHAddress] == nil {
-			contractsAddrs[model.ETHAddress] = make(map[gethCommon.Address]struct{})
+		if e.IsMinerRewardEvent() {
+			// only one event on miner reward
+			totalMinerReward.SetString(e.Value, 10)
+		} else if e.IsUncleRewardEvent() {
+			// it's two uncle events possible.
+			d, ok := new(big.Int).SetString(e.Value, 10)
```
Comment (markya0616) [inline store/transfer_processor.go]: I think we can ignore the `ok` check and make code clear

File: store/transfer_processor.go [hunk 3]
```diff
@@ -63,21 +64,16 @@ func newTransferProcessor(blockNumber int64,
...
+	totalUncleRewards := new(big.Int)
 
 	// Collect modified addresses
 	seenAddrs := make(map[gethCommon.Address]struct{})
 	var addrs [][]byte
-	// From block / uncles coinbases
-	for _, cb := range coinbases {
-		if _, ok := seenAddrs[cb]; !ok {
-			seenAddrs[cb] = struct{}{}
-			addrs = append(addrs, cb.Bytes())
```
Comment (markya0616) [inline store/transfer_processor.go]: We need to keep it. We used `addrs` to query subscriptions and the coinbases may be our subscriptions.

File: store/transfer_processor.go [hunk 4]
```diff
@@ -169,17 +167,19 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
-	// Make sure coinbases are also included in balance queries to geth
-	for _, addr := range coinbases {
-		if allSubs[addr] == nil {
-			continue
-		}
-		if contractsAddrs[model.ETHAddress] == nil {
-			contractsAddrs[model.ETHAddress] = make(map[gethCommon.Address]struct{})
+		if e.IsMinerRewardEvent() {
+			// only one event on miner reward
+			totalMinerReward.SetString(e.Value, 10)
```
Comment (kimmylin) [inline store/transfer_processor.go]: double check we only see miner and uncle reward event one time each per block

File: store/transfer_processor.go (hunk 1 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: ok, but the `if e.IsMinerRewardEvent() || e.IsUncleRewardEvent()` condition expression should be in  `if _, ok := seenAddrs[fromAddr]; !ok`.

File: store/transfer_processor.go (hunk 2 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: OK

File: store/transfer_processor.go (hunk 3 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: this condition are handled in https://github.com/getamis/eth-indexer/pull/54/files#diff-778fd13deb3062e86beaee58b3afff70R96

File: store/transfer_processor.go (hunk 1 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: oh...I couldn't use `continue` keyword. if expression is satisfied, we will miss `toAddr`.

File: store/transfer_processor.go (hunk 4 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: ok

File: store/transfer_processor.go [hunk 5]
```diff
@@ -295,29 +307,60 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
+			if f, ok := feeDiff[addr]; ok {
 				if totalFees[sub.Group] == nil {
 					totalFees[sub.Group] = new(big.Int).Set(f)
 				} else {
 					totalFees[sub.Group] = new(big.Int).Add(f, totalFees[sub.Group])
 				}
 			}
+			// Consider miner reward
+			if r, ok := minerRewardDiff[addr]; ok {
+				s.logger.Info("add miner reward to total balance")
```
Comment (markya0616) [inline store/transfer_processor.go]: Remove this log

File: store/transfer_processor.go [hunk 6]
```diff
@@ -295,29 +307,60 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
+				if totalMinerReward[sub.Group] == nil {
+					totalMinerReward[sub.Group] = new(big.Int).Set(r)
```
Comment (markya0616) [inline store/transfer_processor.go]: `totalMinerReward[sub.Group]` should only set, and never add?

File: store/transfer_processor.go [hunk 7]
```diff
@@ -169,17 +169,29 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
-			continue
-		}
-		if contractsAddrs[model.ETHAddress] == nil {
-			contractsAddrs[model.ETHAddress] = make(map[gethCommon.Address]struct{})
+		if e.IsMinerRewardEvent() {
+			to := gethCommon.BytesToAddress(e.To)
+			reward, _ := new(big.Int).SetString(e.Value, 10)
+
+			if len(minerRewardDiff) > 1 {
+				s.logger.Error("Detect two miner reward events on block", "block", e.BlockNumber, "coinbase", common.BytesTo0xHex(e.To), "balance", e.Value)
```
Comment (kimmylin) [inline store/transfer_processor.go]: maybe find a way to log the old one too?

File: store/transfer_processor.go [hunk 8]
```diff
@@ -169,17 +169,29 @@ func (s *transferProcessor) process(ctx context.Context, events []*model.Transfe
...
+				return common.ErrTooManyMiners
+			}
+			minerRewardDiff[to] = new(big.Int).Set(reward)
+		} else if e.IsUncleRewardEvent() {
+			to := gethCommon.BytesToAddress(e.To)
+			reward, _ := new(big.Int).SetString(e.Value, 10)
+			if uncleRewardDiff[to] == nil {
+				uncleRewardDiff[to] = new(big.Int).Set(reward)
+			} else {
+				if len(uncleRewardDiff) > model.MaxUncles {
```
Comment (kimmylin) [inline store/transfer_processor.go]: this check should be done around line 183, outside this loop

Comment (kimmylin) [reply store/transfer_processor.go]: also, if you can find a way to log the old ones, it'll be awesome.

File: store/transfer_processor.go (hunk 5 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: 🤣

File: model/types.go [hunk 9]
```diff
@@ -31,9 +31,13 @@ var (
...
+	// RewardToMiner represents a constant at from field in transfer event
+	RewardToMiner = common.BytesToAddress([]byte("MINER REWARD"))
+	// RewardToUncle represents a constant at from field in transfer event
+	RewardToUncle = common.BytesToAddress([]byte("UNCLE REWARD"))
 
 	// Maximum number of uncles allowed in a single block
-	maxUncles         = 2
+	MaxUncles         = 2
 	ErrTooManyUncles  = errors.New("too many uncles or confused numbers of uncle")
 	ErrConfusedUncles = errors.New("confused numbers of uncle")
```
[the remark this answers is not in the document]
Comment (bailantaotao) [reply model/types.go]: i can't...the cycle import is caused by `ErrTooManyUncles`...

File: store/transfer_processor.go (hunk 8 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: OK

File: store/transfer_processor.go (hunk 7 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: add new struct function

```
func (s *transferProcessor) printUnexpectedRewardEvent(e *model.Transfer, rewardDiff map[gethCommon.Address]*big.Int) {
	for addr, value := range rewardDiff {
		s.logger.Error("Previous event diff", "block number", e.BlockNumber, "from", string(e.From), "to", addr.Hex(), "balance", value)
	}

	s.logger.Error("Current event diff", "block number", e.BlockNumber, "from", string(e.From), "to", string(e.To), "balance", e.Value)
}

```

File: store/transfer_processor.go (hunk 8 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: add new struct function
```
func (s *transferProcessor) printUnexpectedRewardEvent(e *model.Transfer, rewardDiff map[gethCommon.Address]*big.Int) {
	for addr, value := range rewardDiff {
		s.logger.Error("Previous event diff", "block number", e.BlockNumber, "from", string(e.From), "to", addr.Hex(), "balance", value)
	}

	s.logger.Error("Current event diff", "block number", e.BlockNumber, "from", string(e.From), "to", string(e.To), "balance", e.Value)
}

```

File: store/transfer_processor.go (hunk 6 above)
Comment (bailantaotao) [reply store/transfer_processor.go]: OK, i remove the `add` function, and return error after event detecting was hit with invalid transfer.  https://github.com/getamis/eth-indexer/pull/54/files#diff-778fd13deb3062e86beaee58b3afff70R177

File: store/transfer_processor.go (hunk 8 above)
Comment (kimmylin) [reply store/transfer_processor.go]: thank you!

**Pull request merged**